switchio/loops/proactor.py

```diff
@@ -110,6 +117,84 @@ def start(self):
         self.log.debug("Exiting listen loop")
         self._running = False
 
+    async def _process_event(self, e, evname):
+        '''Process an ESL event by delegating to the appropriate handler
+        and any succeeding callback chain. This is the core handler lookup
+        routine and should be optimized for speed.
+
+        An event is considered consumed if:
+        1) the handler + callback chain returns True
+        2) the handler + callback chain raises a special exception
+
+        :param ESL.ESLEvent e: event received over esl on self._rx_con
+        :param str evname: event type/name string
+        '''
+        # epoch is the time when first event is received
+        if self._epoch:
+            self._fs_time = get_event_time(e)
+        else:
+            self._epoch = self._fs_time = get_event_time(e)
+
+        consumed = False  # is this event consumed by a handler/callback
+        if 'CUSTOM' in evname:
+            evname = e.get('Event-Subclass')
+        self.log.debug("receive event '{}'".format(evname))
+
+        uid = e.get('Unique-ID')
+
+        handler = self._handlers.get(evname, False)
+        loop = self.loop
+        if handler:
+            self.log.debug("handler is '{}'".format(handler))
+            try:
+                consumed, ret = utils.uncons(*handler(e))  # invoke handler
+                model = ret[0]
+
+                # attempt to lookup a consuming client app (callbacks) by id
+                cid = model.cid if model else self.get_id(e, 'default')
+                self.log.debug("consumers id is '{}'".format(cid))
+
+                if model:
+                    # signal any awaiting futures
+                    fut = model._futures.get(evname, None)
+                    if fut:
+                        fut.set_result(e)
+                        # seriously guys, this is literally so stupid
+                        await asyncio.sleep(0)  # resume waiting coroutines...
```

Comment:
@vodik you ever look at the [stdlib's impl](https://github.com/python/cpython/blob/master/Lib/asyncio/tasks.py#L476) for this?
Seems to me there should be some explicit way to resume all coroutines waiting on a task/future.

Do you guys have a more proper way in `aiosip`?

Reply:
Heh so looking into this apparently it's a common quiff with `asyncio`.
This [SO answer](https://stackoverflow.com/questions/29868372/python-asyncio-run-event-loop-once/29868627#29868627) claims the correct solution is to stop and start the loop...